app/controllers/TemplateController.php

```diff
@@ -1,18 +1,93 @@
 <?php
 
-use Phalcon\Mvc\Model\Resultset;
+use Phalcon\Db;
+use Phalcon\Paginator\Adapter\NativeArray as PaginatorArray;
 
 class TemplateController extends BaseController
 {
     public function listAction()
     {
-        $categoryArray = Category::find(
+        // Fetch URL params
+        $categoryId     = $this->dispatcher->getParam("category");
+        $currentPageNr    = $this->dispatcher->getParam("page");
+
+        // Find all categories for sidebar
+        $categories = Category::find(
             array(
                 'columns'   => 'id, name, parent_id',
                 'order'     => 'name ASC'
             )
         )->toArray();
 
-        $this->view->setVar("categoryArray",    $categoryArray);
+        $templates = $this->getTemplatesForCategory($categoryId);
+
+        // Initialize pagination
+        $paginator = new PaginatorArray(
```

Comment:
Use QueryBuilder as data source, native array tends to get very slow if there is a large number of entries.